tests/testthat/test-equals_na_linter.R

```diff
@@ -1,45 +1,59 @@
-test_that("returns the correct linting", {
+test_that("equals_na_linter skips allowed usages", {
   linter <- equals_na_linter()
-  msg <- rex::rex("Use is.na for comparisons to NA (not == or !=)")
+
   expect_lint("blah", NULL, linter)
   expect_lint("  blah", NULL, linter)
   expect_lint("  blah", NULL, linter)
   expect_lint("x=NA", NULL, linter)
+  expect_lint("x = NaN", NULL, linter)
+  expect_lint("x = NA_real_", NULL, linter)
+  expect_lint("is.na(x)", NULL, linter)
+  expect_lint("is.nan(x)", NULL, linter)
+  expect_lint("x[!is.na(x)]", NULL, linter)
 
-  expect_lint(
-    "x == NA",
-    list(message = msg, line_number = 1L, column_number = 1L),
-    linter
-  )
+  # equals_na_linter should ignore strings and comments
+  expect_lint("is.na(x) # dont flag x == NA if inside a comment", NULL, linter)
+  expect_lint("msg <- 'dont flag x == NA if inside a string'", NULL, linter)
 
-  expect_lint(
-    "x==NA",
-    list(message = msg, line_number = 1L, column_number = 1L),
-    linter
-  )
+  # nested NAs are okay
+  expect_lint("x==f(1, ignore = NA)", NULL, linter)
+})
 
+skip_if_not_installed("tibble")
+skip_if_not_installed("patrick")
+patrick::with_parameters_test_that(
+  "equals_na_linter blocks disallowed usages for all combinations of operators and types of NAs",
   expect_lint(
-    "x==f(1, ignore = NA)",
-    NULL,
-    linter
+    sprintf("x %s %s", operation, type_na),
+    rex::rex("Use is.na for comparisons to NA (not == or !=)"),
+    equals_na_linter()
+  ),
+  .cases = tibble::tribble(
+    ~.test_name,                ~operation, ~type_na,
```

Comment:
here's a case where `expand.grid()` is really better but `.test_name` is annoying. any ideas?

Reply:
Something like this, you mean?

``` r
op <- c("==", "!=")
op_verbose <- c("equality", "inequality")
type_na <- c("NA", "NA_real_", "NA_complex_")

expand.grid(
  .test_name = paste(op_verbose, type_na, sep = ", "),
  operation = op,
  type_na = type_na,
  stringsAsFactors = FALSE
) |>
  tibble::as_tibble()
#> # A tibble: 18 × 3
#>    .test_name            operation type_na    
#>    <chr>                 <chr>     <chr>      
#>  1 equality, NA          ==        NA         
#>  2 inequality, NA_real_  ==        NA         
#>  3 equality, NA_complex_ ==        NA         
#>  4 equality, NA          !=        NA         
#>  5 inequality, NA_real_  !=        NA         
#>  6 equality, NA_complex_ !=        NA         
#>  7 equality, NA          ==        NA_real_   
#>  8 inequality, NA_real_  ==        NA_real_   
#>  9 equality, NA_complex_ ==        NA_real_   
#> 10 equality, NA          !=        NA_real_   
#> 11 inequality, NA_real_  !=        NA_real_   
#> 12 equality, NA_complex_ !=        NA_real_   
#> 13 equality, NA          ==        NA_complex_
#> 14 inequality, NA_real_  ==        NA_complex_
#> 15 equality, NA_complex_ ==        NA_complex_
#> 16 equality, NA          !=        NA_complex_
#> 17 inequality, NA_real_  !=        NA_complex_
#> 18 equality, NA_complex_ !=        NA_complex_
```

<sup>Created on 2022-10-08 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
that one's incorrect IINM :)

see https://github.com/google/patrick/issues/9, maybe we use the `rownames_to_column()` approach for now?

Reply:
punting for now...